modules/apps/depot/depot-test/src/testIntegration/java/com/liferay/depot/service/test/DepotEntryDLAppServiceWhenCopyingWithDocumentTypesTest.java

```diff
@@ -0,0 +1,406 @@
+			).build(),
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), "description"
+			).build(),
+			ServiceContextTestUtil.getServiceContext());
+
+		_depotGroup = _groupLocalService.getGroup(_depotEntry.getGroupId());
+
+		DDMStructure depotDDMStructure = _ddmStructureLocalService.addStructure(
+			_depotGroup.getCreatorUserId(), _depotGroup.getGroupId(),
+			DDMStructureConstants.DEFAULT_PARENT_STRUCTURE_ID,
+			PortalUtil.getClassNameId(DLFileEntryMetadata.class),
+			StringPool.BLANK,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(),
+				DLFileEntryMetadata.class.getSimpleName()
+			).build(),
+			new HashMap<>(), StringPool.BLANK, StorageType.DEFAULT.toString(),
+			ServiceContextTestUtil.getServiceContext(_depotGroup.getGroupId()));
+
+		_depotDLFileEntryType = _dlFileEntryTypeLocalService.addFileEntryType(
+			_depotGroup.getCreatorUserId(), _depotGroup.getGroupId(),
+			depotDDMStructure.getStructureId(), null,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), RandomTestUtil.randomString()
+			).build(),
+			new HashMap<>(),
+			DLFileEntryTypeConstants.FILE_ENTRY_TYPE_SCOPE_DEFAULT,
+			ServiceContextTestUtil.getServiceContext(_depotGroup.getGroupId()));
+
+		_group = GroupTestUtil.addGroup();
+
+		DDMStructure ddmStructure = _ddmStructureLocalService.addStructure(
+			_group.getCreatorUserId(), _group.getGroupId(),
+			DDMStructureConstants.DEFAULT_PARENT_STRUCTURE_ID,
+			PortalUtil.getClassNameId(DLFileEntryMetadata.class),
+			StringPool.BLANK,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(),
+				DLFileEntryMetadata.class.getSimpleName()
+			).build(),
+			new HashMap<>(), StringPool.BLANK, StorageType.DEFAULT.toString(),
+			ServiceContextTestUtil.getServiceContext(_group.getGroupId()));
+
+		_dlFileEntryType = _dlFileEntryTypeLocalService.addFileEntryType(
+			_group.getCreatorUserId(), _group.getGroupId(),
+			ddmStructure.getStructureId(), null,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), RandomTestUtil.randomString()
+			).build(),
+			new HashMap<>(),
+			DLFileEntryTypeConstants.FILE_ENTRY_TYPE_SCOPE_DEFAULT,
+			ServiceContextTestUtil.getServiceContext(_group.getGroupId()));
+	}
+
+	@Test(expected = InvalidFileEntryTypeException.class)
+	public void testCopyFileEntryFailsUnlessDLFileEntryTypeAvailable()
+		throws Exception {
+
+		_depotEntryGroupRelLocalService.addDepotEntryGroupRel(
+			true, _depotEntry.getDepotEntryId(), _group.getGroupId(), false);
+
+		ServiceContext serviceContext =
+			ServiceContextTestUtil.getServiceContext(_group.getGroupId());
+
+		serviceContext.setAttribute(
+			"fileEntryTypeId", _dlFileEntryType.getFileEntryTypeId());
+
+		FileEntry fileEntry1 = _dlAppService.addFileEntry(
+			RandomTestUtil.randomString(), _group.getGroupId(),
+			DLFolderConstants.DEFAULT_PARENT_FOLDER_ID, _FILE_NAME,
+			ContentTypes.TEXT_PLAIN, _FILE_NAME, StringPool.BLANK,
+			StringPool.BLANK, StringPool.BLANK, new byte[0], null, null,
+			serviceContext);
+
+		_dlAppService.copyFileEntry(
```

Comment:
Similarly, I would create a method `_copyFileEntry(long fileEntryId, long fileEntryTypeId, Group group)`

With these private methods is also clearer what are you changing in every call. Currently the all look the same.

Reply:
Do you think is worth doing this refactoring?

Reply:
I've created the method in one of the tests only. In the other one, the differences make it not worth (we end up with a private method with almost as many parameters as the one in the service).